### PR TITLE
Add patch to support multi vector in faiss

### PR DESCRIPTION
**Pull request by heemin32**
### Description
Add patch to support multi vector in faiss

Introduced an extension point in the way of collecting search result in HNSW algorithm. `ResultCollectorFactory` is passed as part of a search parameter. For each search inside faiss library, it create a new `ResultCollector` using the factory. The `ResultCollector`'s `collect` method is used to collect search data.

This enable us to pass custom ResultCollector with which we can dedupe vector on its parent Id so that exact k result of original document can be returned for nested field.
 
* PR for patch file to make it easy to review. https://github.com/heemin32/faiss/pull/1/files
### Issues Resolved
https://github.com/opensearch-project/k-NN/issues/1065
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by ryanbogan — DISMISSED** [review]
LGTM!

File: .github/workflows/CI.yml
Comment (jmazanec15) [inline .github/workflows/CI.yml:41]: What is meant by this? If we do this, everytime we add a patch we will need to update this.

File: .github/workflows/CI.yml
Comment (jmazanec15) [inline .github/workflows/CI.yml:48]: Why remove?

File: .github/workflows/CI.yml
Comment (jmazanec15) [inline .github/workflows/CI.yml:46]: Can we add comment explaining this?

File: CHANGELOG.md
Comment (jmazanec15) [inline CHANGELOG.md:31]: Is this a refactor or a feature?

File: jni/CMakeLists.txt
Comment (jmazanec15) [inline jni/CMakeLists.txt:89]: Why is this needed?

File: jni/CMakeLists.txt
Comment (jmazanec15) [inline jni/CMakeLists.txt:158]: Can we loop over patches?

File: jni/patches/faiss/0001-Custom-patch-to-support-multi-vector.patch
Comment (jmazanec15) [inline jni/patches/faiss/0001-Custom-patch-to-support-multi-vector.patch:1]: Can we add a README about patches? Or update DEV guide? Want to make sure we know what we are doing on lib upgrades.

**Comment by jmazanec15** [comment]
Im wondering if for patches we can raise a PR against our personal forks of faiss and link them in the PR so we can review the code changes as opposed to review via git diff. Not sure if there is a better approach. What do you think @heemin32 ?

File: .github/workflows/CI.yml
Comment (heemin32) [reply .github/workflows/CI.yml:41]: Good Idea. Let me do that.

**Comment by heemin32** [comment]
> Im wondering if for patches we can raise a PR against our personal forks of faiss and link them in the PR so we can review the code changes as opposed to review via git diff. Not sure if there is a better approach. What do you think @heemin32 ?

Good Idea. Let me do that.

File: .github/workflows/CI.yml
Comment (heemin32) [reply .github/workflows/CI.yml:41]: Correct. Git command in CMAKE file like `git submodule update --init --` or `git apply` does not work for ubuntu. Need to talk with @peterzhuamazon to see if we can fix it by using different ubuntu image.

File: .github/workflows/CI.yml
Comment (heemin32) [reply .github/workflows/CI.yml:48]: Commented in line #42. `# Deleting file at the end to skip `git apply` inside CMAKE file`

File: CHANGELOG.md
Comment (heemin32) [reply CHANGELOG.md:31]: I put it as refactoring because this PR itself does not provide any new feature to users. However, I added CHANGELOG because it might have a impact in performance of current feature and we want to catch the root cause easily if that happen.

File: jni/CMakeLists.txt
Comment (heemin32) [reply jni/CMakeLists.txt:89]: In mac, I have faiss in `/opt/homebrew/share` and the command fail to pull faiss submodule. I don't have nmslib but thought that there is no harm on doing same for nmslib.

File: jni/CMakeLists.txt
Comment (heemin32) [reply jni/CMakeLists.txt:158]: There won't be many patches I hope. Also, explicit applying would be better to make sure that we are applying all intended patches.

File: jni/patches/faiss/0001-Custom-patch-to-support-multi-vector.patch
Comment (heemin32) [reply jni/patches/faiss/0001-Custom-patch-to-support-multi-vector.patch:1]: Let me add in DEV guide on how to generate patches or how to update patches.

File: CHANGELOG.md
Comment (navneet1v) [reply CHANGELOG.md:31]: @heemin32 what I would say lets create a feature branch for this change and add all the changes in that feature branch. Then apply a single commit in main branch as part of 1 PR. This will ensure that this feature is packed in 1 PR. This is similar to what I did for Faiss Filtering also. It can avoid comments like this.

Comment (heemin32) [reply CHANGELOG.md:31]: Changed to feature branch.

**Comment by heemin32** [comment]
Test for patch completed successfully. https://github.com/facebookresearch/faiss/pull/3178

File: .github/workflows/CI.yml
Comment (navneet1v) [reply .github/workflows/CI.yml:41]: @heemin32 what was the conclusion for this?

File: jni/CMakeLists.txt
Comment (navneet1v) [reply jni/CMakeLists.txt:151]: +1 we should apply patches explicitly.

**Comment by navneet1v** [comment]
Overall code looks good to me. Please resolve all the comments before I can approve the PR

File: .github/workflows/CI.yml
Comment (heemin32) [reply .github/workflows/CI.yml:41]: No update yet. Need to talk with @peterzhuamazon once he come back from a vacation. The code is needed as of now to make it work though.

Comment (navneet1v) [reply .github/workflows/CI.yml:41]: is that discussion blocking this PR?

Comment (heemin32) [reply .github/workflows/CI.yml:41]: It won't.

**Review by jmazanec15 — APPROVED** [review]

**Review by navneet1v — APPROVED** [review]

**Pull request merged**
